Datashader/Numba various LoweringErrors

- [x] I am using `'0.50.0dev0+71.gce2139c'` installed from the numba channel on anaconda.org
- [x] I have included below a minimal working reproducer

Steps to reproduce:

* Setup a development environment with Datashader `v0.11.0a1` and the Numba version listed above.
* Run `pytest datashader/tests/test_raster.py`

As far as I can tell, errors all seem to be of the form:

```
E       numba.core.errors.LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
E       Buffer dtype cannot be buffer, have dtype: array(float32, 2d, C)
E       
E       File "datashader/resampling.py", line 901:
E       def _downsample_2d_std_var(src, mask, use_mask, method, fill_value,
E           <source elided>
E       
E           for out_y in prange(out_h):
E           ^
E       
```

Log file is attached.

[datashader_numba_error.log](https://github.com/numba/numba/files/4485685/datashader_numba_error.log)

--- a/numba/parfors/parfor.py
+++ b/numba/parfors/parfor.py
@@ -3194,14 +3194,16 @@ def _combine_params_races_for_ssa_names(scope, params, races):
             # XXX: it's a bug that something references an undefined name
             return k
 
-    keep = set(races)
+    races1 = set(races)
+    unver_params = list(map(unversion, params))
+
     for rv in races:
-        if any(unversion(rv) == unversion(pv) for pv in params):
-            keep.discard(rv)
+        if any(unversion(rv) == pv for pv in unver_params):
+            races1.discard(rv)
         else:
             break
 
-    return params | keep, keep
+    return params | races1, races1
 
 
 def get_parfor_params_inner(parfor, pre_defs, options_fusion, fusion_info):

--- a/numba/parfors/parfor.py
+++ b/numba/parfors/parfor.py
@@ -3169,12 +3169,39 @@ def get_parfor_params(blocks, options_fusion, fusion_info):
             dummy_block.body = block.body[:i]
             before_defs = compute_use_defs({0: dummy_block}).defmap[0]
             pre_defs |= before_defs
-            parfor.params = get_parfor_params_inner(parfor, pre_defs, options_fusion, fusion_info) | parfor.races
+            params = get_parfor_params_inner(
+                parfor, pre_defs, options_fusion, fusion_info,
+            )
+            parfor.params, parfor.races = _combine_params_races_for_ssa_names(
+                block.scope, params, parfor.races,
+            )
             parfor_ids.add(parfor.id)
             parfors.append(parfor)
 
         pre_defs |= all_defs[label]
     return parfor_ids, parfors
+
+
+def _combine_params_races_for_ssa_names(scope, params, races):
+    """Returns `(params|races1, races1)`, where `races1` contains all variables
+    in `races` are NOT referring to the same unversioned (SSA) variables in
+    `params`.
+    """
+    def unversion(k):
+        try:
+            return scope.get_exact(k).unversioned_name
+        except ir.NotDefinedError:
+            # XXX: it's a bug that something references an undefined name
+            return k
+
+    keep = set(races)
+    for rv in races:
+        if any(unversion(rv) == unversion(pv) for pv in params):
+            keep.discard(rv)
+        else:
+            break
+
+    return params | keep, keep
 
 
 def get_parfor_params_inner(parfor, pre_defs, options_fusion, fusion_info):

--- a/numba/tests/test_parfors.py
+++ b/numba/tests/test_parfors.py
@@ -3428,6 +3428,26 @@ class TestParforsMisc(TestParforsBase):
 
         self.check(parallel_test, size, arr)
 
+    @skip_parfors_unsupported
+    def test_issue5570_ssa_races(self):
+        @njit(parallel=True)
+        def foo(src, method, out):
+            for i in prange(1):
+                for j in range(1):
+                    out[i, j] = 1
+            if method:
+                out += 1
+            return out
+
+        src = np.zeros((5,5))
+        method = 57
+        out = np.zeros((2, 2))
+
+        self.assertPreciseEqual(
+            foo(src, method, out),
+            foo.py_func(src, method, out)
+        )
+
 
 @skip_parfors_unsupported
 class TestParforsDiagnostics(TestParforsBase):

